run.py

```diff
@@ -452,8 +452,9 @@ def fetch_test_details(var) -> dict:
                 res.append(tmp)
```

Comment:
We are discovering the tests again further down. It would be nice if we do it once... I like the idea of discovering the tests before the execution `setup | run | teardown `

Reply:
Will handle this in another PR.